specs/SharedBuffer.md

```diff
@@ -0,0 +1,333 @@
+
+When the application code calls `PostSharedBufferToScript`, the script side will
+receive a `SharedBufferReceived` event containing the buffer as an `ArrayBuffer` object.
+After receiving the shared buffer object, it can access it the same way as any other
+ArrayBuffer object, including transering to a web worker to process the data on
+the worker thread.
+
+As shared buffer normally represent a large memory, instead of waiting for garbage
+collection to release the memory along with the owning objects, the application
+should try to release the buffer as soon as it doesn't need access to it.
+This can be done by calling Close() method on the shared buffer object, or
+`chrome.webview.releaseBuffer` from script.
+
+As the memory could contain sensitive information and corrupted memory could crash
+the application, the application should only share buffer with trusted sites.
+
+The application can use other messaging channel like `PostWebMessageAsJson` and
+`chrome.webview.postMessage` to inform the other side the desire to have a shared
+buffer and the status of the buffer (like data is produced or consumed).
+
+# Examples
+
+The example below illustrates how to send data from application to script for one time read only consumption.
+
+The script code will look like this:
+```
+        window.onload = function () {
+            window.chrome.webview.addEventListener("SharedBufferReceived", e => {
+                SharedBufferReceived(e);
+            });
+        }
+
+        function SharedBufferReceived(e) {
+            if (e.data && e.data.readOnly) {
+                // This is the one time read only buffer
+                let oneTimeSharedBuffer = e.sharedBuffer;
+                // Consume the data from the buffer
+                DisplaySharedBufferData(oneTimeSharedBuffer);
+                // Release the buffer after consuming the data.
+                chrome.webview.releaseBuffer(oneTimeSharedBuffer);
+            }
+        }
+```
+## Win32 C++
+```cpp
+
+        wil::com_ptr<ICoreWebView2SharedBuffer> sharedBuffer;
+        CHECK_FAILURE(webviewEnvironment->CreateSharedBuffer(bufferSize, &sharedBuffer));
+        // Fill data into the shared memory via IStream.
+        wil::com_ptr<IStream> stream;
+        CHECK_FAILURE(sharedBuffer->GetStream(&stream));
+        CHECK_FAILURE(stream->Write(data, dataSize, nullptr));
+        PCWSTR additionalDataAsJson = L"{\"readOnly\":true}";
+        if (forFrame)
+        {
+            m_webviewFrame->PostSharedBufferToScript(
+                sharedBuffer.get(), /*isReadOnlyToScript*/TRUE, additionalDataAsJson);
+        }
+        else
+        {
+            m_webView->PostSharedBufferToScript(
+               sharedBuffer.get(), /*isReadOnlyToScript*/TRUE, additionalDataAsJson);
+        }
+        // Explicitly close the one time shared buffer to ensure that the resource is released timely.
+        sharedBuffer->Close();
```

Comment:
It does not. The implementation takes care of locking the underlying resources until it is posted.

Reply:
Ok, so the sample here needs to show that with a comment, like:

> The buffer automatically releases any resources it may have allocated when the last reference to the buffer is released by both the app and any script in the web view. For deterministic release of the section use the `Close()` method, but only when your app is certain the view has finished use of it.
>
> After the `Close()` method is called, any calls to `get_Size`, `get_Buffer`, or `GetStream` will return E_ABORT. The pointer returned from `get_Buffer` will no longer be valid. Read, write, and seek operations on the `IStream` returned from `GetStream` will also return E_ABORT.

Reply:
The memory can be written/read multiple times after it is posted to script. PostSharedBufferToScript basically means share the memory with script. After it is shared, each side can access it freely, until Close is called and the after that point, the native code side cannot access the memory anymore. Close() means that this process will not access the memory anymore.